happytransformer/trainers/trainer_qa.py

```diff
@@ -0,0 +1,192 @@
+                correct += 1
+            count += 1
+
+            if output_filepath != None:
+                results.append(
+                    {
+                        "contexts": context,
+                        "questions": question,
+                        "answer": answer["answer_text"].lower(),
+                        "outputs": output,
+                        "correct": output == compare_answer
+                    }
+                )
+
+            self._print_status(init_time, count, total, update_interval, correct/count)
+
+        score = correct/total
+        ending = str(round(score, 2) * 100) + "%"
+
+        result_output = "Evaluating Result: " + str(correct) + "/" + str(total) + " -- " + ending
+        self.logger.info(result_output)
+
+        if output_filepath != None:
+            fieldnames = ["contexts", "questions", "answer", "outputs", "correct"]
+            self._output_result_to_csv(output_filepath, fieldnames, results)
+
+        return score
+
+
+    def test(self, filepath, args, output_filepath):
+        #todo
+        pass
+
+    def __get_train_eval_data(self, filepath):
+        """
+        Used for parsing data for training and evaluating (both contain labels)
+        :param filepath: a string that contains the location of the data
+        :return:
+        """
+
+        contexts = []
+        questions = []
+        answers = []
+        with open(filepath, newline='') as csv_file:
+            reader = csv.DictReader(csv_file)
+            for row in reader:
+                contexts.append(row['context'])
+                questions.append(row['question'])
+                answer = {}
+                answer["answer_text"] = row['answer_text']
+                answer["answer_start"] = int(row['answer_start'])
+                answers.append(answer)
+        csv_file.close()
+
+        return contexts, questions, answers
+
+
+    def __add_end_idx(self, contexts, answers):
+        for answer, context in zip(answers, contexts):
+
+            gold_text = answer['answer_text']
+            start_idx = answer['answer_start']
+            end_idx = start_idx + len(gold_text)
+
+            # todo (maybe): strip answer['text'] (remove white space from start and end)
+            # sometimes squad answers are off by a character or two – fix this
+            if context[start_idx:end_idx] == gold_text:
+                answer['answer_end'] = end_idx
+            elif context[start_idx - 1:end_idx - 1] == gold_text:
+                answer['answer_start'] = start_idx - 1
+                answer['answer_end'] = end_idx - 1  # When the gold label is off by one character
+            elif context[start_idx - 2:end_idx - 2] == gold_text:
+                answer['answer_start'] = start_idx - 2
+                answer['answer_end'] = end_idx - 2  # When the gold label is off by two characters
+            else:
+                print("error: implement skipping training answer")
```

Comment:
could you link a SQUAD example where this edge case is necessary?

Reply:
The official documentation does not indicate which examples cause this edge case. It would be a good experiment tho.

Reply:
"sometimes squad answers are off by a character or two" is there a source for this?

Reply:
Source: https://huggingface.co/transformers/custom_datasets.html#question-answering-with-squad-2-0

<img width="1295" alt="Screen Shot 2021-01-02 at 12 28 29 AM" src="https://user-images.githubusercontent.com/38766185/103451380-85947500-4c91-11eb-8df3-c65a6307d526.png">

Reply:
https://gist.github.com/ted537/29a656d7bbce7247a3c89b7d4c011015

I created a script to test for this issue and it didn't come up. I believe it only occurs if you save the SQUAD JSON with the wrong encoding. This occurs because converting to ASCII can reduce certain UTF8 characters away. We can probably add a warning to users in the case where we have to clean up the indices.

TLDR: The SQUAD dataset is correct, but becomes incorrect if saved in the wrong encoding (user error quite possible)